### PR TITLE
chore: Lint duplicate manifests

### PR DESCRIPTION
**Pull request by Kaitou786**
Signed-off-by: Tarun Khandelwal <tarun@flanksource.com>

### Description

Check all the manifests and see if there is any duplicate keys found the identification key is in the form `<kind>-<name>-<namespace>` or `<kind>-<name>` in case namespace is not present

### Dependencies

- _Ex. Other PRs._

### Breaking Change

- [ ] Yes
- [x] No

### Testing Notes

**What I did:**
Just run the linter file directly `go run test/linter/main.go`

**How you can replicate my testing:**
_How can the reviewer validate this PR?_

### **Links**

_Issues links or other related resources_

### CODE REVIEW

File: test/linter/main.go [hunk 1]
```diff
@@ -17,8 +19,19 @@ const (
 	configFile = "test/linter.yaml"
 )
 
+var allowedDuplicateKeys = []string{"CustomResourceDefinition-servicemonitors.monitoring.coreos.com"}
+var ignoreManifestsSubPaths = []string{
```
Comment (moshloop) [inline test/linter/main.go:23]: only the upstream should be skipped

**Comment by BrendanGalloway** [comment]
@Kaitou786 would you mind putting in a line or two for description in the cover sheet, so that reviewers have a bit of context as to what the change is intended to do.

File: test/linter/main.go [hunk 2]
```diff
@@ -49,6 +54,57 @@ func checkPath(filePath string, info os.FileInfo, err error) error {
...
+	}
+	manifestData, err := kommons.GetUnstructuredObjects(yamlFile)
+	if err != nil {
+		log.Warnf("error parsing the yaml %v", manifest)
+	}
+	for i := range manifestData {
+		if manifestData[i].Object["metadata"].(map[string]interface{})["namespace"] != nil {
+			value := fmt.Sprintf("%v-%v-%v", manifestData[i].Object["kind"].(string), manifestData[i].Object["metadata"].(map[string]interface{})["name"].(string), manifestData[i].Object["metadata"].(map[string]interface{})["namespace"].(string))
+			if !contains(allowedDuplicateKeys, value) && contains(keys, value) {
+				return errors.Errorf("error %v already present in the manifests", value)
```
Comment (BrendanGalloway) [inline test/linter/main.go:94]: Could the error also state which file the duplicate is encountered in?

**Comment by Kaitou786** [comment]
> @Kaitou786 would you mind putting in a line or two for description in the cover sheet, so that reviewers have a bit of context as to what the change is intended to do.

Ah sure I will from now on.

File: test/linter/main.go (hunk 2 above)
Comment (Kaitou786) [reply test/linter/main.go:94]: Thinking of a way it could be possible.

Comment (Kaitou786) [reply test/linter/main.go]: Changed keys to a map with map-key as the generated key and value as the manifest file. You can test by edit the allowedDuplicateManifests var to any other value and see the test fails with giving msg like:
```
FATA[0000] Manifest linting failed: error: key CustomResourceDefinition-servicemonitors.monitoring.coreos.com from manifests/crd/service-monitor.yaml already present in manifests/crd/prometheus.yaml
```

**Review by BrendanGalloway — APPROVED** [review]

File: test/linter/main.go (hunk 2 above)
Comment (BrendanGalloway) [reply test/linter/main.go]: Thanks, that is a nice level of verbosity

**Pull request merged**